### PR TITLE
Meta: Explain how to build with ninja

### PR DESCRIPTION
**Pull request by BenWiederhake**
Inspired by #3047, and my struggles to understand how cmake is supposed to work ^^

CC @Dexesttp, because he wrote the original description, and probably knows better commands for day-to-day use.

### CODE REVIEW

File: Documentation/BuildInstructions.md [hunk 1]
```diff
@@ -107,7 +107,16 @@ Outside of QEMU, Serenity will run on VirtualBox. If you're curious, see how to
...
-You may also want to replace `make` with `ninja` in the above commands for some additional build speed benefits. To do this, go to an empty directory at the root and call `cmake .. -G Ninja` inside that directory. You might either create a new directory or reuse the existing `Build` directory after cleaning it.
+#### Faster than make: "Ninja"
+
+You may also want to replace `make` with `ninja` in the above commands for some additional build speed benefits, like reduced double-building of headers.
+Most of the process stays the same:
+- Go to an empty directory at the root (e.g. `Build/`) and call `cmake .. -G Ninja` inside that directory
+- You might either create a new directory or reuse the existing `Build` directory after cleaning it.
+- `make` becomes `cmake --build .` (or possibly `cmake --build . --verbose -j 4`)
+- `make install` becomes `cmake --install .` (or possibly `cmake --install . --strip`, if you know what you're doing)
+- `make image` becomes `cmake --build . -- image`
```
Comment (bgianfo) [inline Documentation/BuildInstructions.md:118]: Don't all the targets exist for ninja as well? 
- ninja 
- ninja install
- ninja image
- ninja run

Comment (BenWiederhake) [reply Documentation/BuildInstructions.md:118]: Ah, for some reason I thought "ninja probably needs some extra weird arguments", looked at how `./BuildIt.sh` works, found `cmake --build . --target LibC`, and then it was all downhill from there ^^'

I'll simplify it then.

Comment (BenWiederhake) [reply Documentation/BuildInstructions.md]: Done.

**Comment by Dexesttp** [comment]
Hmm. Yeah, I think this wasn't clear enough in the original documentation, sorry ^^'

`cmake` is not a build tool. It is a tool that generates config files for build tools, but you cannot build anything with `cmake` directly (just like you cannot compile with `make` directly, the compiler will be `gcc` or `clang`).

By default, when you run `cmake ..`, it generates Makefile config files for your build system (equivalent to running `cmake .. -G Makefile`). Then, once your Makefile files are created properly, you just use the `make` command to use them, and `make` will call `gcc` and so on. It works just like if you wrote your Makefile yourself, except you didn't have to.

The alternative (and advantage of cmake) is that you can decide to generate Ninja config files instead of Makefile config files (by running the `cmake .. -G Ninja` command). Once you execute the command, the Ninja config files are read when you use the `ninja` command, and `ninja` will call `gcc` and so on. Same thing, you could have written your Ninja config files yourself, but you didn't have to (and Ninja files are harder to write by yourself than Makefile files).

The thing is that, since both the Makefile and Ninja config files are generated from the same "guidelines", it doesn't matter what tool you actually use, they will have the same generated subcommands. So, from the user's point of view, `ninja` becomes a drop-in replacement of  `make`.

This means that all the commands you are used to writing with `make` can be used identically with `ninja`, just with `s/make/ninja/` :
- `make` becomes `ninja`
- `make install` becomes `ninja install`
- `make image` becomes `ninja image`
- `make run` becomes `ninja run`
- `make my-custom-target --with-custom-options` becomes `ninja my-custom-target --with-custom-options`

And so on :)

**Comment by BenWiederhake** [comment]
Thanks for all the feedback! I also included the thing about `-j`, that's really nice! :^)

**Comment by selectiveduplicate** [comment]
Thanks for this; to me it's now become much easier to understand how to replace `make` with `ninja`.

**Comment by BenWiederhake** [comment]
In case you actually mean it: Thanks, it was a journey for me, too ^^

In case this was sarcasm: Sure, if one already knows how to use `ninja`, then the explanation doesn't add any value. However, the `BuildInstructions.md` are, like all kinds of documentation, for the people who don't know yet how Stuff™ works. And as you can see from my initial attempts, my guesses about ninja were far from reality, and a short quip that `make == ninja` would have saved me a few minutes. So I *do* believe this change is an improvement.

**Comment by selectiveduplicate** [comment]
> In case you actually mean it: Thanks, it was a journey for me, too ^^
> 
> In case this was sarcasm: Sure, if one already knows how to use `ninja`, then the explanation doesn't add any value. However, the `BuildInstructions.md` are, like all kinds of documentation, for the people who don't know yet how Stuff™ works. And as you can see from my initial attempts, my guesses about ninja were far from reality, and a short quip that `make == ninja` would have saved me a few minutes. So I _do_ believe this change is an improvement.

It wasn't sarcasm! I hardly know enough about build systems, let alone `ninja`; honestly this change has made the documentation better and clearer.

**Comment by Dexesttp** [comment]
Well, it definitely feels like this documentation change is worth merging then !

@awesomekling any comments on the proposed updated documentation for ninja ?

**Pull request merged**